### ISSUE TITLE
Can useMutation be called asynchronously?

### ISSUE THREAD

**Issue by lucasmogari** [tannerlinsley/react-query#241]
When making a lot of requests, I noted that a request is only made, only after the previous on returned.

So if a request is slow to return, the calls to the mutate function are ignored while waiting.

Is this how it's supposed to work?

**Comment by tannerlinsley** [comment]
So it’s seems less to do with the asynchronous nature and more to do with denouncing. Now that you bring it up there are definitely use cases for firing off multiple mutations in succession. This would mean though that previous promises would be ignored in the status, data and error state tracking.

**Comment by lucasmogari** [comment]
I was using for a chat app. Every time an user sends a message, it would send a request to save the message. So some messages were ignored. Maybe this is not a good use case for useMutation.

**Comment by tannerlinsley** [comment]
`useMutation` is really useful primarily for a few things:

- Simply tracking promise status and data/error
- Declaratively using `onSuccess/onError/onSettled` to trigger query refetching or direct cache mutations

If you don't need or want either of those, then useMutation may not be useful to you.  However, if you are interested in either of those, we could easily add in what I talked about where `status`, `data` and `error` would only be tracked for the latest issued promise. The onWhatever callbacks would still get fired for all issued mutations though.

**Comment by lucasmogari** [comment]
So I'll use it for another things later.

But for this chat, the useInfiniteQuery is awesome for listing messages/scrolling old messages!

**Comment by tannerlinsley** [comment]
Cool. Also, FYI, but the latest version allows `useMutation` to have multiple outgoing requests.

**Comment by GeospatialMax** [comment]
@tannerlinsley I am running into an issue where I call the mutate function returned by `useMutation` multiple times and it only ever seems to run the mutation for the very last call. Is this expected behaviour? And is there any way to get around it? Your last post suggests that it should be possible to have mutliple consecutive calls, but how would I achieve this?

Thanks in advance!

**Comment by tannerlinsley** [comment]
I'll take a looksie.

**Comment by tannerlinsley** [comment]
You should be able to `mutate()` as much as you want, but the lifecycle options will only work for the latest call you've made.

**Comment by tannerlinsley** [comment]
So if you're firing off multiple mutations that overlap, you can't rely on lifecycle callbacks getting called all the time for out-of-date mutations that were already issued.

**Comment by kizivat** [comment]
Hello, just came across the same issue. I created a sandbox replicating the issue here: https://codesandbox.io/s/awesome-williams-gt0ib?file=/src/App.js

**Comment by tannerlinsley** [comment]
What version are y’all on?

**Comment by kizivat** [comment]
> What version are y’all on?

The sandbox above uses version 1.5.8

**Comment by GeospatialMax** [comment]
> What version are y’all on?

I am on 1.5.5 at the moment.

**Comment by kizivat** [comment]
Jus tried it in 2.0.9, seems to have the same issue

**Comment by dan-dr** [comment]
I've also stumbled on a use case where I have many buttons that use the same mutation with different parameters (e.g Add friend button in a list of people), and I need to track each mutation status. Is it possible at all with v1/v2?

**Comment by tannerlinsley** [comment]
Here's a sandbox: https://codesandbox.io/s/unruffled-sid-g2tzx?file=/src/App.js

- The mutation function will call it as many time as you hit the button.
- The `onMutate` callback will fire for every time
- The `onSuccess` (and other delayed callbacks) will only fire for the latest mutation

**Comment by wtanna** [comment]
I should have looked here before I got too deep! I also created a sandbox of the issue.

https://codesandbox.io/s/kind-wood-kz89k?file=/pages/index.js

This is causing some pain points for us on slower connections when a user is updating multiple things at once. We have an expensive query to fetch all the data initially. Therefore, we leverage `useMutation` and `onSuccess` to update the `queryCache` directly with the returned response from our API. 

However, like the sandboxes represent, only the last result is being returned which makes our UI be in a "stuck" state as it will just render loading spinners forever. Unless the user leaves the page and comes back.

Should we open up a new issue? Or is there a work around for this?

**Comment by wtanna** [comment]
I took a deeper look into this and found that versions <1.5.4 work. Version 1.5.5 had this change which causes the issue:

https://github.com/tannerlinsley/react-query/commit/4f56c31ba5bf898805fba4f48479ebc6a2b93ff6

When I had the `async` back in my local `node_modules` our apps work fine again. Not sure what kind of ramifications adding it back in will have though at this moment.

**Comment by tannerlinsley** [comment]
Sooo.... here's the dealio. The `isLatest()` guards were put there on purpose, so that only the latest mutation would have it's side-effects called. I can see that this may be problematic though. I think what I'm hearing is that you want all of the mutation side effects to fire no matter what, but this will probably introduce race conditions where one mutation could complete before another. Do we want to enforce side-effect order with the order that the mutation was called so side effects are stable?

- The easiest way to avoid race conditions is to only use the "latest" issued mutation, which is how it is currently
- The next easiest way to avoid race conditions, but ensure that all side-effects fire is to chain the side-effects based on their order. This means that if you fire off 2 commands for `mutate(a)` and `mutate(b)`, the side effects for `a` would have to finish before the side effects for `b` are run.

**Comment by tannerlinsley** [comment]
This would essentially do what you want. Each mutate would have side-effects, but only the latest mutation would be allowed update the mutation state:

```js
const mutate = React.useCallback(
    async (
      variables,
      { onSuccess = noop, onError = noop, onSettled = noop, throwOnError } = {}
    ) => {
      const config = getConfig()

      const mutationId = uid()
      latestMutationRef.current = mutationId

      const isLatest = () => latestMutationRef.current === mutationId

      let snapshotValue

      try {
        dispatch({ type: actionLoading })
        snapshotValue = await config.onMutate(variables)

        let data = await getMutationFn()(variables)

        if (isLatest()) {
          dispatch({ type: actionResolve, data })
        }

        await config.onSuccess(data, variables)
        await onSuccess(data, variables)
        await config.onSettled(data, null, variables)
        await onSettled(data, null, variables)

        return data
      } catch (error) {
        Console.error(error)
        await config.onError(error, variables, snapshotValue)
        await onError(error, variables, snapshotValue)
        await config.onSettled(undefined, error, variables, snapshotValue)
        await onSettled(undefined, error, variables, snapshotValue)

        if (isLatest()) {
          dispatch({ type: actionReject, error })
        }

        if (throwOnError ?? config.throwOnError) {
          throw error
        }
      }
    },
    [dispatch, getConfig, getMutationFn]
  )
```

**Comment by qkreltms** [comment]
https://github.com/tannerlinsley/react-query/issues/241#issuecomment-648973545

> Here's a sandbox: https://codesandbox.io/s/unruffled-sid-g2tzx?file=/src/App.js
> 
> * The mutation function will call it as many time as you hit the button.
> * The `onMutate` callback will fire for every time
> * The `onSuccess` (and other delayed callbacks) will only fire for the latest mutation

Can you describe it on document page?
